### PR TITLE
Merging in New Pipeline

### PR DESCRIPTION
**Pull request by ASAAR**
We now have a new working pipeline that can be run by going to modules/hub/ and running hub.py. It has all of the features of the old system, and more. There is still a lot of work to be done however.

This PR closes,
- Adding the Hub #115 
- Adding the Worker #112 
- Adding the Gym #113 
- Adding Model Filtering #114 
- Adding the model controller #87 
- Adding the data controller #86 

All of these work together and do not break current tests. They have not replaced the old system entirely, but they will soon.

### CODE REVIEW

**Comment by asclines** [comment]
What parts of the old system does this _not_ replace?

**Comment by ASAAR** [comment]
@asclines I mean, this doesn't have Ryan's stuff integrated yet. But we need to wait until we have some strategy patterns to govern that.

**Comment by ASAAR** [comment]
@RyanMcBerg @ZakeryFyke @telelu03 No one going to call me out on not adding tests?

**Comment by ZakeryFyke** [comment]
Hey @ASAAR you seem to lack test cases for this newest pull request you've done.

**Comment by ASAAR** [comment]
@ZakeryFyke Thanks! I almost forgot about that ;)

File: modules/DMZ/data_kit/data_splitting.py [hunk 1]
```diff
@@ -29,14 +34,14 @@ def get_train_test(pandas_data, target_col):
...
     # Selection of training/target data for validation and training.
-    data = pd.DataFrame.as_matrix(pandas_data)
-    target_loc = pandas_data.columns.get_loc(target_col)
-    y = data[:, target_loc]
-
     x = pandas_data.drop(target_col, 1)
-    x = pd.DataFrame.as_matrix(x)
+    y = pandas_data[[target_col]]
 
     return x, y
```
Comment (asclines) [inline modules/DMZ/data_kit/data_splitting.py]: x,y are poor variable names for this. Maybe something like predictors and response, or predictors and targets

File: modules/DMZ/data_kit/validation_package.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
+from modules.DMZ.data_kit import data_splitting
+
+
+class ValidationPackage(object):
```
Comment (asclines) [inline modules/DMZ/data_kit/validation_package.py]: Needs docs. What is this supposed to validate? etc..

File: modules/DMZ/models/model_filter.py [hunk 3]
```diff
@@ -1,29 +1,20 @@
...
+    modules = sn.get_modules(sn.path_to_models(prediction_type))
 
-    for filename in __all__:
+    for filename in modules:
         importer = __import__(filename)
         for function in dir(importer):
-            if function.__str__()[0:5] == "train":
-                models.add(function)
+            item = getattr(importer, function)
+            if function.__str__()[0:5] == "train" and callable(item):
```
Comment (asclines) [inline modules/DMZ/models/model_filter.py]: Is this the best way to find out if valid?

File: modules/DMZ/ticket.py [hunk 4]
```diff
@@ -5,10 +5,14 @@
...
         self.training = train
         self.testing = test
         self.target = label
-        self.training_data_in_folder = train_folder
-        self.testing_data_in_folder = test_folder
+        self.hyperparameters = None
+        self.validation_results = None
+        self.test_results = None
+        self.parameters_used = None
+        self.max_iterations = 0
```
Comment (asclines) [inline modules/DMZ/ticket.py]: Maybe also add something about current iteration?

File: modules/DMZ/utils/config_options.py [hunk 5]
```diff
@@ -1,25 +1,26 @@
...
         self.target_column = raw_input("Provide the name of the column to be predicted: ")
 
-    #Check input from user
+    # Check input from user
     def validate_options(self):
         predict = self.prediction_type.lower()
-        if predict != "regression" and predict != "classification" and predict != "auto":
+        if predict != "regression" and predict != "classification":
             print("Please input the type of prediction to be done.")
             return
```
Comment (asclines) [inline modules/DMZ/utils/config_options.py]: Soo.. what happens here? It seem's like the program just stalls but doesn't end?

File: modules/DMZ/utils/system_navigation.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
+from os.path import dirname, basename, isfile
+import glob
+import setup
+import sys
+
+
+def path_to_models(prediction_type):
```
Comment (asclines) [inline modules/DMZ/utils/system_navigation.py]: Suggestion: Add some check to make sure this path actually exists & is valid.

File: modules/DMZ/utils/system_navigation.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+    root_path = setup.get_root_path()
+    return root_path + "/modules/DMZ/models/" + prediction_type + '/'
+
+
+def add_path_to(path):
```
Comment (asclines) [inline modules/DMZ/utils/system_navigation.py]: Method name doesn't make sense

File: modules/DMZ/utils/system_navigation.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    sys.path.append(path)
+
+
+def get_modules(path_to_modules):
+    modules = glob.glob(dirname(path_to_modules) + "/*.py")
+    return [basename(f)[:-3] for f in modules if isfile(f)]
```
Comment (asclines) [inline modules/DMZ/utils/system_navigation.py]: Yes it's cool to do as much in one line as possible but will make hard to maintain in future.

File: modules/hub/README.md [hunk 9]
```diff
@@ -0,0 +1,5 @@
+# Overview
+The hub should only do work that is necessary for the deployment of Workers,
```
Comment (asclines) [inline modules/hub/README.md]: Does this need to be in the README? The README holds end-user docs. The end-user doesn't care about this.

File: modules/toolbox/ml_runners.py [hunk 10]
```diff
@@ -1,13 +1,9 @@
...
+import sys
+
 sys.path.insert(0, os.path.abspath('../..'))
 
-import setup
-from validation_package import ValidationPackage
-from modules.DMZ.data_kit import data_io
-from modules.DMZ.models.regression import scikit_online_regressors
-from modules.DMZ.models.regression import scikit_regression_learners
 from modules.toolbox import *
```
Comment (asclines) [inline modules/toolbox/ml_runners.py]: why this way?

File: modules/worker/data_controller.py [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+from modules.DMZ.data_kit import data_scaling
+from modules.DMZ.data_kit import data_splitting
+
+
+# This class will manage the data for the worker and prep it for use.
+# Eventually it will make changes to the data. Currently it does not.
+class DataController(object):
+
+    def __init__(self, training, testing, target):
+        # self.validation_pack = validation_package.ValidationPackage().prepare_package(training, target)
```
Comment (asclines) [inline modules/worker/data_controller.py]: Should this be here?

File: modules/DMZ/data_kit/data_splitting.py (hunk 1 above)
Comment (asclines) [inline modules/DMZ/data_kit/data_splitting.py]: Okay, after reviewing more, I noticed the consistency of using `x` and `y` as variable names so I could be talked out of thinking that.

File: modules/DMZ/data_kit/data_filler.py [hunk 12]
```diff
@@ -2,35 +2,40 @@
 import numpy as np
 
 
-#Drops rows which have fewer than threshold observations.
-#e.g. if threshold = 5, any rows with fewer than 5 observations will be removed from pandas_data.
+# Drops rows which have fewer than threshold observations.
+# e.g. if threshold = 5, any rows with fewer than 5 observations will be removed from pandas_data.
```
Comment (ASAAR) [inline modules/DMZ/data_kit/data_filler.py]: Can you clarify what you mean? The README at this level with this file? Also, part of this documentation does exist in that README.

Comment (asclines) [inline modules/DMZ/data_kit/data_filler.py]: Good question; We should figure out how we want the end-user docs look. I think at this level would be find as technically the end-users of this file will be developers of this project so they will go this deep. Honestly, maybe don't put too much thought into _where_ right now. Let's just all make sure we get documentation written and as we finalize the structure of this project we can move the docs to the right spot.

File: modules/worker/worker.py [hunk 13]
```diff
@@ -14,10 +14,16 @@ def __init__(self, hub):
...
+    # in order to train a specific machine learning algorithm.
     def setup_controllers(self):
         self.data_control = data_controller.DataController(self.ticket.training, self.ticket.testing, self.ticket.target)
         self.model_control = model_controller.ModelController(self.ticket.ml_algorithm)
         self.model_gym = gym.Gym()
 
+    # Runs the model using the prepped data. Validation results will be saved
+    # with MLTA, results from the trained model applied to test data will be returned
+    # and used on their own or by another service.
+    # TODO: Integrate test data application.
```
Comment (asclines) [inline modules/worker/worker.py]: Who will be responsible for completing this TODO? Is there an issue for it?

File: modules/DMZ/data_kit/data_filler.py (hunk 12 above)
Comment (ASAAR) [inline modules/DMZ/data_kit/data_filler.py]: So I guess what extra documentation do you think we need? An example of calling the function with example data that shows how it is used? I want to document, but I also don't want to over-document.

Comment (asclines) [inline modules/DMZ/data_kit/data_filler.py]: Over-documentation is a concern I'm seeing too. We do want to show at least function input/outputs for all external methods.

**Comment by ASAAR** [comment]
@asclines @ZakeryFyke Sorry, I haven't had time to write tests yet, that'll happen tonight

**Comment by ZakeryFyke** [comment]
@ASAAR We'll let it slide this time

**Pull request merged**